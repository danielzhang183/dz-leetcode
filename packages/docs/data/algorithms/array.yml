questions:
  - name: Two Sum
    title: two-sum
    difficulty: Easy
    id: '1'
    link: /algorithms/array/001
    origin: https://leetcode.cn/problems/two-sum
    tag: array
    category: algorithms
