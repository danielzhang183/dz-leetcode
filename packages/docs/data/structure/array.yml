questions:
  - name: 两数之和
    id: 46
    link: /array/
    rank: easy
  - name: 删除有序数组中的重复项
    id: 26
    link: /array/026
    rank: easy
  - name: 移除元素
    id: 27
    link: /array/027
    rank: easy
  - name: 移动零
    id: 283
    link: /array/283
    rank: easy
  - name: 有序数组的平方
    id: 977
    link: /array/977
    rank: easy
  - name: 长度最小的子数组
    id: 209
    link: /array/209
    rank: easy
  - name: 反转字符串
    id: 344
    link: /array/344
    rank: easy
  - name: 两数之和-ii-输入有序数组
    id: 167
    link: /array/167
    rank: easy
  - name: 验证回文串
    id: 125
    link: /array/125
    rank: easy
  - name: 两个数组的交集
    id: 349
    link: /array/
    rank: easy
