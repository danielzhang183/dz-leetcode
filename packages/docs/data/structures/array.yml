description: A linear data structure with contiguous memory.
questions:
  - name: Two Sum
    title: two-sum
    difficulty: Easy
    id: 1
    link: /structures/array/001
    origin: https://leetcode.cn/problems/two-sum
    tag: array
    category: structures
    done: true
  - name: Remove Duplicates from Sorted Array
    title: remove-duplicates-from-sorted-array
    difficulty: Easy
    id: 26
    link: /structures/array/026
    origin: https://leetcode.cn/problems/remove-duplicates-from-sorted-array
    tag: array
    category: structures
    done: true
  - name: Remove Element
    title: remove-element
    difficulty: Easy
    id: 27
    link: /structures/array/027
    origin: https://leetcode.cn/problems/remove-element
    tag: array
    category: structures
    done: true
  - name: Find First and Last Position of Element in Sorted Array
    title: find-first-and-last-position-of-element-in-sorted-array
    difficulty: Medium
    id: 34
    link: /structures/array/034
    origin: >-
      https://leetcode.cn/problems/find-first-and-last-position-of-element-in-sorted-array
    tag: array
    category: structures
  - name: Search Insert Position
    title: search-insert-position
    difficulty: Easy
    id: 35
    link: /structures/array/035
    origin: https://leetcode.cn/problems/search-insert-position
    tag: array
    tags:
      - binary-search
    category: structures
    done: true
  - name: Valid Palindrome
    title: valid-palindrome
    difficulty: Easy
    id: 125
    link: /structures/array/125
    origin: https://leetcode.cn/problems/valid-palindrome
    tag: array
    category: structures
    done: true
  - name: Two Sum II - Input Array Is Sorted
    title: two-sum-ii-input-array-is-sorted
    difficulty: Medium
    id: 167
    link: /structures/array/167
    origin: https://leetcode.cn/problems/two-sum-ii-input-array-is-sorted
    tag: array
    category: structures
    done: true
  - name: Rotate Array
    title: rotate-array
    difficulty: Medium
    id: 189
    link: /structures/array/189
    origin: https://leetcode.cn/problems/rotate-array
    tag: array
    category: structures
  - name: Minimum Size Subarray Sum
    title: minimum-size-subarray-sum
    difficulty: Medium
    id: 209
    link: /structures/array/209
    origin: https://leetcode.cn/problems/minimum-size-subarray-sum
    tag: array
    category: structures
    done: true
  - name: Move Zeroes
    title: move-zeroes
    difficulty: Easy
    id: 283
    link: /structures/array/283
    origin: https://leetcode.cn/problems/move-zeroes
    tag: array
    tags:
      - two-pointers
    category: structures
    done: true
  - name: Reverse String
    title: reverse-string
    difficulty: Easy
    id: 344
    link: /structures/array/344
    origin: https://leetcode.cn/problems/reverse-string
    tag: array
    tags:
      - two-pointers
      - api
    category: structures
    done: true
  - name: Intersection of Two Arrays
    title: intersection-of-two-arrays
    difficulty: Easy
    id: 349
    link: /structures/array/349
    origin: https://leetcode.cn/problems/intersection-of-two-arrays
    tag: array
    tags:
      - set
      - sort + two-pointers
    category: structures
    done: true
  - name: Find Pivot Index
    title: find-pivot-index
    difficulty: Easy
    id: 724
    link: /structures/array/724
    origin: https://leetcode.cn/problems/find-pivot-index
    tag: array
    tags:
      - prefix-sum
    category: structures
    done: true
  - name: Sort Array By Parity II
    title: sort-array-by-parity-ii
    difficulty: Easy
    id: 922
    link: /structures/array/922
    origin: https://leetcode.cn/problems/sort-array-by-parity-ii
    tag: array
    category: structures
  - name: Valid Mountain Array
    title: valid-mountain-array
    difficulty: Easy
    id: 941
    link: /structures/array/941
    origin: https://leetcode.cn/problems/valid-mountain-array
    tag: array
    category: structures
  - name: Squares of a Sorted Array
    title: squares-of-a-sorted-array
    difficulty: Easy
    id: 977
    link: /structures/array/977
    origin: https://leetcode.cn/problems/squares-of-a-sorted-array
    tag: array
    category: structures
  - name: Unique Number of Occurrences
    title: unique-number-of-occurrences
    difficulty: Easy
    id: 1207
    link: /structures/array/1207
    origin: https://leetcode.cn/problems/unique-number-of-occurrences
    tag: array
    category: structures
  - name: How Many Numbers Are Smaller Than the Current Number
    title: how-many-numbers-are-smaller-than-the-current-number
    difficulty: Easy
    id: 1365
    link: /structures/array/1365
    origin: >-
      https://leetcode.cn/problems/how-many-numbers-are-smaller-than-the-current-number
    tag: array
    category: structures
