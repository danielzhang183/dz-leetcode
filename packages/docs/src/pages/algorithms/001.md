---
title: Two Sum
display: 1. Two Sum
difficulty: Easy
origin: https://leetcode.cn/problems/two-sum
---

[[toc]]

## Problem

Given an array of integers nums and an integer target, return <em>indices of the two numbers such that they add up to target</em>.

You may assume that each input would have <strong><em>exactly</em> one solution</strong>, and you may not use the <em>same</em> element twice.

You can return the answer in any order.

&nbsp;
<strong>Example 1:</strong>

<pre>
<strong>Input:</strong> nums = [2,7,11,15], target = 9
<strong>Output:</strong> [0,1]
<strong>Explanation:</strong> Because nums[0] + nums[1] == 9, we return [0, 1].
</pre>

<strong>Example 2:</strong>

<pre>
<strong>Input:</strong> nums = [3,2,4], target = 6
<strong>Output:</strong> [1,2]
</pre>

<strong>Example 3:</strong>

<pre>
<strong>Input:</strong> nums = [3,3], target = 6
<strong>Output:</strong> [0,1]
</pre>

&nbsp;
<strong>Constraints:</strong>

<ul>
  <li>2 &lt;= nums.length &lt;= 10<sup>4</sup></li>
  <li>-10<sup>9</sup> &lt;= nums[i] &lt;= 10<sup>9</sup></li>
  <li>-10<sup>9</sup> &lt;= target &lt;= 10<sup>9</sup></li>
  <li><strong>Only one valid answer exists.</strong></li>
</ul>

&nbsp;
<strong>Follow-up:&nbsp;</strong>Can you come up with an algorithm that is less than&nbsp;O(n<sup>2</sup>)&nbsp;time complexity?

## Solution

```ts
export function twoSum(nums: number[], target: number): number[] {

};
```

[view source](https://leetcode.cn/problems/two-sum)
