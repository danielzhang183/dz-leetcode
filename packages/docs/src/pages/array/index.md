---
title: Array
display: ''
plum: true
---

<SubNav/>

<ListPosts />
