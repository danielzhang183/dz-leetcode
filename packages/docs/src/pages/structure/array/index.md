---
title: Array
display: Array
subtitle: List of questions that I am aware of
description: List of questions that I am aware of
plum: true
---

<SubNav module="structure" />

<ListQuestions module="structure" tag="array" />
