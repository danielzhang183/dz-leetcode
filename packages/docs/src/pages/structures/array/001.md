---
title: Two Sum
display: 1. Two Sum
difficulty: Easy
origin: https://leetcode.cn/problems/two-sum
---

[[toc]]

## Problem

Given an array of integers nums and an integer target, return `indices of the two numbers such that they add up to target`.

You may assume that each input would have **`exactly` one solution**, and you may not use the `same` element twice.

You can return the answer in any order.

## Usage

### Example 1

```
input: nums = [2,7,11,15], target = 9
output: [0,1]
Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
```

### Example 2

```
input: nums = [3,2,4], target = 6
output: [1,2]
```

### Example 3

```
input: nums = [3,3], target = 6
output: [0,1]
```


**Constraints:**

- <code>2 &lt;= nums.length &lt;= 10<sup>4</sup></code>
- <code>-10<sup>9</sup> &lt;= nums[i] &lt;= 10<sup>9</sup></code>
- <code>-10<sup>9</sup> &lt;= target &lt;= 10<sup>9</sup></code>
- **Only one valid answer exists.**


**Follow-up:** Can you come up with an algorithm that is less than O(n<sup>2</sup>) time complexity?

## Solution

```ts
function twoSum(nums: number[], target: number): number[] {

}
```

[view source](https://leetcode.cn/problems/two-sum)
