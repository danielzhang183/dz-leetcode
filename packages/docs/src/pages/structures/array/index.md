---
title: Array
titleZh: 数组
subtitle: The most basic yet important data structure is the array. It is a linear data structure.
subtitleZh: 最基本但最重要的数据结构，它是一种线性数据结构。
plum: true
---

<SubNav module="structures" />

An array is a collection of `homogeneous` data types where the elements are allocated `contiguous` memory. 
 
Because of the contiguous allocation of memory, any element of an array can *be accessed in constant time*. Each array element has a corresponding index number.

<ListQuestions module="structures" tag="array" />
