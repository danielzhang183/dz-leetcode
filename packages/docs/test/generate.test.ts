import { describe, expect, it } from 'vitest'
import { genCatelog, genCode, genMarkdown, genTestCase } from '../scripts'
import { resolvedQuestion } from './fixture'

describe('generate content', () => {
  it('markdown', async () => {
    const markdown = await genMarkdown(resolvedQuestion)
    expect(markdown).toMatchSnapshot('markdown')
    const { outFile } = markdown
    expect(outFile).toMatchInlineSnapshot('"/home/t04857/packages/docs/src/pages/algorithms/array/001.md"')
  })

  it('code', async () => {
    const code = await genCode(resolvedQuestion)
    expect(code).toMatchSnapshot('code')
    const { outFile, content } = code
    expect(content).toMatchInlineSnapshot(`
      "export function twoSum(nums: number[], target: number): number[] {

      };"
    `)
    expect(outFile).toMatchInlineSnapshot('"/home/t04857/packages/code/src/algorithms/array/001.ts"')
  })

  it('testcases', async () => {
    const testcases = await genTestCase(resolvedQuestion)
    expect(testcases).toMatchSnapshot('testcases')
    const { outFile, content } = testcases
    expect(content).toMatchInlineSnapshot(`
      "import { describe, expect, it } from 'vitest'
      import { twoSum } from '../../../src/algorithms/array/001'

      describe('twoSum', () => {
        it('exported', () => {
          expect(twoSum([2,7,11,15])).toBe(9)
          expect(twoSum([3,2,4])).toBe(6)
          expect(twoSum([3,3])).toBe(6)
        })
      })
      "
    `)
    expect(outFile).toMatchInlineSnapshot('"/home/t04857/packages/code/test/algorithms/array/001.test.ts"')
  })

  it('catelog', async () => {
    const catelog = await genCatelog(resolvedQuestion)
    expect(catelog).toMatchSnapshot('catelog')
    const { outFile, content } = catelog
    expect(content).toMatchInlineSnapshot(`
      "questions:
        - name: Two Sum
          title: two-sum
          difficulty: Easy
          id: 1
          link: /algorithms/array/001
          origin: https://leetcode.cn/problems/two-sum
          tag: array
          category: algorithms"
    `)
    expect(outFile).toMatchInlineSnapshot('"/home/t04857/packages/docs/data/algorithms/array.yml"')
  })
})
