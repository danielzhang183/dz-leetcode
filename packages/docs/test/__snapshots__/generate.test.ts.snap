// Vitest Snapshot v1

exports[`generate content > catelog > catelog 1`] = `
{
  "content": "questions:
  - name: Two Sum
    title: two-sum
    difficulty: Easy
    id: 1
    link: /algorithms/array/001
    origin: https://leetcode.cn/problems/two-sum
    tag: array
    category: algorithms",
  "outFile": "/home/t04857/packages/docs/data/algorithms/array.yml",
  "type": "category",
}
`;

exports[`generate content > code > code 1`] = `
{
  "content": "export function twoSum(nums: number[], target: number): number[] {

};",
  "outFile": "/home/t04857/packages/code/src/algorithms/array/001.ts",
  "type": "code",
}
`;

exports[`generate content > markdown > markdown 1`] = `
{
  "content": "---
title: Two Sum
display: 1. Two Sum
difficulty: Easy
origin: https://leetcode.cn/problems/two-sum
---

[[toc]]

## Problem

Given an array of integers nums and an integer target, return <em>indices of the two numbers such that they add up to target</em>.

You may assume that each input would have <strong><em>exactly</em> one solution</strong>, and you may not use the <em>same</em> element twice.

You can return the answer in any order.

 
<strong>Example 1:</strong>

<pre>
<strong>Input:</strong> nums = [2,7,11,15], target = 9
<strong>Output:</strong> [0,1]
<strong>Explanation:</strong> Because nums[0] + nums[1] == 9, we return [0, 1].
</pre>

<strong>Example 2:</strong>

<pre>
<strong>Input:</strong> nums = [3,2,4], target = 6
<strong>Output:</strong> [1,2]
</pre>

<strong>Example 3:</strong>

<pre>
<strong>Input:</strong> nums = [3,3], target = 6
<strong>Output:</strong> [0,1]
</pre>

 
<strong>Constraints:</strong>

<ul>
  <li>2 &lt;= nums.length &lt;= 10<sup>4</sup></li>
  <li>-10<sup>9</sup> &lt;= nums[i] &lt;= 10<sup>9</sup></li>
  <li>-10<sup>9</sup> &lt;= target &lt;= 10<sup>9</sup></li>
  <li><strong>Only one valid answer exists.</strong></li>
</ul>

 
<strong>Follow-up: </strong>Can you come up with an algorithm that is less than O(n<sup>2</sup>) time complexity?

## Solution

\`\`\`ts
function twoSum(nums: number[], target: number): number[] {

};
\`\`\`

[view source](https://leetcode.cn/problems/two-sum)
",
  "outFile": "/home/t04857/packages/docs/src/pages/algorithms/array/001.md",
  "type": "markdown",
}
`;

exports[`generate content > testcases > testcases 1`] = `
{
  "content": "import { describe, expect, it } from 'vitest'
import { twoSum } from '../../../src/algorithms/array/001'

describe('twoSum', () => {
  it('exported', () => {
    expect(twoSum([2,7,11,15])).toBe(9)
    expect(twoSum([3,2,4])).toBe(6)
    expect(twoSum([3,3])).toBe(6)
  })
})
",
  "outFile": "/home/t04857/packages/code/test/algorithms/array/001.test.ts",
  "type": "testcase",
}
`;
